---
title: "xena_data_analysis"
format: html
editor: visual
---

## Load packages and targets

```{r}
library(tidyverse)
library(R.utils)
```

```{r}
target_list <- c("PD-L1", "PDL1", "PDL2", "PD-L2", "IDO1", "IDO-1", "TDO", "ARG1", "ARG-1", "ARG2", "ARG-2", "TGFB1", "CCL22", "MCP1", "LGALS3", "GARP","LRRC32", "IL10", "SIGLEC15", "MLIAP")

```

## Download data

```{r}
# Cohort:
cohort = "TCGA_Acute_Myeloid_Leukemia_LAML"
file_path <- "C:/Users/Mette/OneDrive - Danmarks Tekniske Universitet/11. Semester Speciale/Data/Xena/"
dir.create(file_path, showWarnings = FALSE, recursive = TRUE)
```

```{r}
## Data
data_type = "gene_expression_RNAseq_pancan_normalized"

# Specify the URL of the file you want to download
url <- "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.LAML.sampleMap%2FHiSeqV2_PANCAN.gz"

# Specify the file name and location where you want to save the file on your computer
file_name_data <- paste0(cohort, "_", data_type)

if (!(file.exists(paste(file_path, file_name_data, ".tsv.gz", sep = "")))){
  # Call the download.file() function, passing in the URL and file name/location as arguments
  download.file(url, paste(file_path, file_name_data, ".tsv.gz", sep = ""), mode = "wb")
}

if (!(file.exists(paste(file_path, file_name_data, ".tsv", sep = "")))){
  gunzip(paste(file_path, file_name_data, ".tsv.gz", sep = ""))
}
```

```{r}
file.exists(paste(file_path, file_name_data, ".tsv", sep = ""))
```

```{r}
## Phenotype
data_type = "phenotype"

# Specify the URL of the file you want to download
url <- "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.LAML.sampleMap%2FLAML_clinicalMatrix"

# Specify the file name and location where you want to save the file on your computer
file_name_pheno <- paste0(cohort, "_", data_type)

# Call the download.file() function, passing in the URL and file name/location as arguments
if (!(file.exists(paste(file_path, file_name_pheno, ".tsv", sep = "")))){
  download.file(url, paste(file_path, file_name_pheno, ".tsv", sep = ""), mode = "wb")
}
```

```{r}
df_data <- read.csv(paste(file_path, file_name_data, ".tsv", sep = ""),sep = "\t")
df_phenotype <- read.csv(paste(file_path, file_name_pheno, ".tsv", sep = ""),sep = "\t")
```

## Download gene expression across cohorts

```{r}
file_path <- "C:/Users/Mette/OneDrive - Danmarks Tekniske Universitet/11. Semester Speciale/Data/Xena/"
dir.create(file_path, showWarnings = FALSE, recursive = TRUE)

data_type = "gene_expression_RNAseq_pancan_normalized"

url_list <- c("https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.LAML.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.CHOL.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.CESC.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.DLBC.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.HNSC.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.BRCA.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.PAAD.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.SKCM.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.LUAD.sampleMap%2FHiSeqV2_PANCAN.gz",
              "https://tcga-xena-hub.s3.us-east-1.amazonaws.com/download/TCGA.LUSC.sampleMap%2FHiSeqV2_PANCAN.gz")
cohort_list <- c("TCGA_LAML",
                 "TCGA_CHOL",
                 "TCGA_CESC",
                 "TCGA_DLBC",
                 "TCGA_HNSC",
                 "TCGA_BRCA",
                 "TCGA_PAAD",
                 "TCGA_SKCM",
                 "TCGA_LUAD",
                 "TCGA_LUSC")
```

```{r}
df_data_list <- c()

for (i in 1:length(url_list)){
  url <- url_list[i]
  file_name_data <- paste0(cohort_list[i], "_", data_type)
  
  if (!(file.exists(paste(file_path, file_name_data, ".tsv.gz", sep = "")) | 
        file.exists(paste(file_path, file_name_data, ".tsv", sep = "")))){
    download.file(url, paste(file_path, file_name_data, ".tsv.gz", sep = ""), mode = "wb")
  }
  
  if (!(file.exists(paste(file_path, file_name_data, ".tsv", sep = "")))){
    gunzip(paste(file_path, file_name_data, ".tsv.gz", sep = ""))
  }
  
  df_data_list[[i]] <- read.csv(paste(file_path, file_name_data, ".tsv", sep = ""),sep = "\t")
}

```

```{r}
# Preprocess data frames to be pivot long and add cancer type as column:
preproc_df_list <- Map(function(df, cohort_x) {
  filter(df, sample %in% target_list) %>% 
    rename(target = sample) %>% 
    pivot_longer(cols = -target, 
                 names_to = "sample", 
                 values_to = "gene_exp") %>% 
    mutate(cancer_type = cohort_x)
}, df_data_list, cohort_list)

full_df <- bind_rows(preproc_df_list) 
```

```{r}
targets_in_df_list <- full_df %>% 
  select(target) %>% 
  unique() %>% 
  as.list() %>% 
  unlist()

full_df %>% 
  filter(target == targets_in_df_list[1]) %>% 
  ggplot(aes(x=cancer_type, y=gene_exp)) +
    geom_boxplot() +
    theme_minimal() +
    labs(
      title = paste(targets_in_df_list[1]),
      subtitle = "Gene expression RNAseq pan cancer",
      caption = "Source: Xena browser",
      x = "Cancer type",
      y = "Normalized pan cancer gene exp"
    ) + 
    theme(
      plot.title = element_text(color = "#0099f8", size = 16, face = "bold", hjust = 0.5),
      plot.subtitle = element_text(face = "bold.italic", hjust = 0.5),
      plot.caption = element_text(face = "italic")
    )
```

```{r}

func_preproc_df <- function (df, cohort_x){
  filter(df, sample %in% target_list) %>% 
    rename(target = sample) %>% 
    pivot_longer(cols=-target,
                 names_to="sample",
                 values_to="gene_exp") %>% 
    mutate(cancer_type = cohort_x)
}

filtered_df_list[[1]] %>% 
  filter(sample %in% target_list) %>% 
  rename(target = sample) %>% 
  pivot_longer(cols=-target,
               names_to="sample",
               values_to="gene_exp") %>% 
  mutate(cancer_type = cohort[1])
```

## Plot

```{r}
df_data %>% 
  filter(sample %in% target_list)
```

```{r}
df_phenotype
```
